src/gluonts/model/predictor.py

```diff
@@ -729,3 +737,26 @@ def predict(self, dataset: Dataset, **kwargs) -> Iterator[Forecast]:
             predictions = trained_pred.predict(local_ds, **kwargs)
             for pred in predictions:
                 yield pred
+
+
+class FallbackPredictor(Predictor):
+    @classmethod
+    def from_predictor(cls, base: Predictor) -> Predictor:
+        raise NotImplementedError
+
+
+def fallback(fallback_cls: Type[FallbackPredictor]):
+    def decorator(predict_item):
+        def fallback_predict(self, item: DataEntry) -> Forecast:
+            try:
+                return predict_item(self, item)
+            except Exception as error:
+                logging.warning(
+                    f"Base predictor didn't return result: {error}"
+                )
+                fallback_predictor = fallback_cls.from_predictor(self)
+                return fallback_predictor.predict_item(item)
+
+        return fallback_predict
+
+    return decorator
```

Comment:
What’s the connection between this (`FallbackPredictor`, `fallback`) and the mechanism added to `MeanPredictor` in #240? It’s a bit involved and not apparent, maybe an example would clarify

Reply:
Here is a small example:

```python
class MyPredictor(RepresentablePredictor):
    ....
    @fallback(MeanPredictor)
    def predict_item(item):
        ...
```

The idea is that `FallbackPredictor`s know to be constructed from other predictors. Essentially, this requires that they only have to required arguments `freq` and `predictions_length`, with all others being optional.

Granted, it's a bit ad-hoc at the moment, but I think I like the overall design approach.

Reply:
I see. But this way the fallback mechanism is hardcoded in the `MyPredictor` type, isn't it? Correct me if I'm wrong: one can also do
```
base_predictor = WhateverPredictor(...)
predictor = MeanPredictor.from_predictor(base_predictor)
```
or something similar? Because that seems more like the example you mentioned in #238.

I mean, ultimately just wrapping a sequence of predictors in a `FallbackPredictor` with its own `predict_item` that applies base predictors in sequence until one succeeds seems simple and flexible (and exactly what you suggest in #238). But maybe this is what's happening here under the decorators magic :-)